design/ceph-config-updates.md

```diff
@@ -0,0 +1,127 @@
+
+#### Required minimum
+All Ceph options in the config file can be set via the command line, so it is tempting to remove the
+need for having a `ceph.conf` in containers at all. In practice, we do not know how to make this an
+actual possibility.
+
+The troublesome situation regards Ceph daemons' ability to discover mons. The daemons must have
+`mon host` specified to find mons which are using non-default ports. It is further undesirable to
+specify this via the `--mon-host` flag because if this parameter is specified statically as a
+container argument and the initial mons fail and are replaced by new mons, the daemon will not know
+how to contact the mons if it is restarted.
+
+The proposed solution to this problem and the only config option which have thus far been deemed
+*required* is the `mon host` global option which will be stored in a Kubernetes ConfigMap so that
+containers always have the most updated version of this information. The single `ceph.conf`
+ConfigMap will be mounted to all containers at `/etc/ceph/ceph.conf`. There is a possibility the
+`mon initial members` global option may also be necessary, as it goes hand-in-hand with `mon host`.
+
+With this one -- or these two -- options set via the config file, all daemons should *require* no
+other config file options set.
+
+#### Convenient minimum
+
+There may be a small set of additional options which are okay to have in the config file for
+convenience if they would otherwise be set via command line for every single daemons.
+
+For example:
+- `log file = /dev/stderr` should be set for all daemons to enforce the container standard of
+  sending logging to stdout/stderr.
+- `mon cluster log file = /dev/stderr` also for reasons above
+- `log stderr prefix = "debug "` should be set for all daemons to differentiate logging from auditing
+  - TODO: should this be set via Ceph centralized config if someone wants to change this? What about Luminous?
+- TODO: did I miss any?
+- `fatal signal handlers = false`
+  **TODO MORE INFORMATION NEEDED** why does Rook set this? It's purpose will determine its config location
+
+#### Additional configs for test environments
+
+Test (especially integration tests) may need to specify `osd pool default size = 1` and
+`osd pool default min size = 1` to support running clusters with only one osd. Test environments
+would have a means of doing this fairly easily using the still-existent config override capability.
+
+### Minimal command line options
+
+The minimal set of command line options is pared down to the settings which inform the daemon where
+to find and/or store data.
+
+Required flags:
+- `--keyring` may necessary to inform daemons where to find their keyrings
+- `--{mon,osd,mgr,mds,rgw}-data-dir` settings exist for all daemons, and these will continue to need
+  set since the locations Rook uses are non-default
+- `--run-dir` will need set on the command line since it is non-default and no longer set in the
+  config file
```

Comment:
it would be nice to use the default location for this inside the container so that an admin that runs bash in teh container can use the 'ceph daemon ....' command.

Reply:
As I understand it from Ceph's `options.cc`, the default run dir for daemons is `/var/run/ceph`, though I don't know if data stored there needs to persist to hosts. And the default data dir for daemons is `/var/lib/ceph/$type/ceph-$id/`, is that all correct?

Reply:
Yes.  /var/run/ceph doesn't need to be (probably shouldn't be) exposed outside the container, and it's usually a tmpfs I think.. it just stores unix domain sockets.

Reply:
IMO we should do something like /var/lib/ceph/$cluster_uuid (host) -> /var/lib/ceph (container) for the monitors.

No other daemons have state that needs to persist on the host.  (OSDs are mounted inside the container by ceph-volume.)

Reply:
**RE: /var/run/ceph:** Ack, and thank you.

-----

**RE: host/container paths:**

Currently Rook stores data on the host as follows where `/var/lib/rook` can be substituted with whatever the `dataDirHostPath` is for a given rook cluster.

- `/var/lib/rook/mon-a` is mon run dir
- `/var/lib/rook/mon-a/data` is mon data dir
- `/var/lib/rook/osd0` is osd run and data dir

If a user wishes to collocate 2 clusters on the same hardware, the user *must* change the `dataDirHostPath` to something that is not `/var/lib/rook`. On the host, following your suggestion, I think that `/var/lib/rook/ceph/$cluster_id` would be a better long-term location given that a user may also have a cockroachdb backend (or other backend) collocated and want to use the same `/var/lib/rook` storage dir.

Once the config rewrite has happened and most of the orchestration code is in the operator only, I think it will be more straightforward for to run an init container that migrates data from the "legacy" `/var/lib/rook/$type$id` host directory to `/var/llib/rook/ceph/$fsid/$type-$id` on the host. My tentative vision would be that upgrades would be automatic still, and this could be done safely, but there could be problems I haven't forseen with moving data around on the host like that.

-----

**RE: osds:**

TBH, I don't know quite enough about the osds to grok what you have said. I can lay down what I think I know, and we can discuss further. I guess I can start off by saying that the OSDs are complicated beasts, and the osd code in Rook is something that has clearly evolved over time. After cleaning up the mons and changing the way ceph handles configs, I would be interested to see if I could simplify the osds by doing an overhaul, but this is an aside.

If I understand what you're saying, the OSDs don't need their Ceph data dir persisted to the host. Currently the data is persisted and looks like this:
```
$> ls -alF /var/lib/rook/osd8/
total 56
drwxr--r-- 1 root root  252 Jan  9 15:40 ./
drwxr-xr-x 1 root root   60 Jan  9 15:40 ../
lrwxrwxrwx 1 root root   58 Jan  9 15:40 block -> /dev/disk/by-partuuid/6cfb2831-bb72-4f03-9f06-2c517166a10b
lrwxrwxrwx 1 root root   58 Jan  9 15:40 block.db -> /dev/disk/by-partuuid/0b21577a-eb1c-4574-ac0d-491d526e7a46
lrwxrwxrwx 1 root root   58 Jan  9 15:40 block.wal -> /dev/disk/by-partuuid/0d974406-06cc-48ed-8bc2-c50797aa9c3c
-rw-r--r-- 1 root root    2 Jan  9 15:40 bluefs
-rw-r--r-- 1 root root   37 Jan  9 15:40 ceph_fsid
-rw-r--r-- 1 root root   37 Jan  9 15:40 fsid
-rw-r--r-- 1 root root   56 Jan  9 15:40 keyring   <-- NOTE: this is not copied/restored
-rw-r--r-- 1 root root    8 Jan  9 15:40 kv_backend
-rw-r--r-- 1 root root   21 Jan  9 15:40 magic
-rw-r--r-- 1 root root    4 Jan  9 15:40 mkfs_done
-rw-r--r-- 1 root root    6 Jan  9 15:40 ready
-rw-r--r-- 1 root root 1459 Jan  9 15:40 rook-ceph.config   <-- NOTE: this is not copied/restored
srwxr-xr-x 1 root root    0 Jan  9 15:40 rook-ceph-osd.8.asok=
drwxr--r-- 1 root root   30 Jan  9 15:40 tmp/     <-- NOTE: this has the `activate.monmap` in it
-rw-r--r-- 1 root root   10 Jan  9 15:40 type
-rw-r--r-- 1 root root    2 Jan  9 15:40 whoami
```

Whenever an OSD is created, Rook makes copies of these files so that it can restore the OSD if the data is lost. For disk-based OSDs, is it not necessary to persist this data?

I know that it must be necessary to persist data to the host when using directory-based OSDs. I can't say I'm a fan of the dir-based osds; they make my developer life hard. But I do understand the desire to have them for testing, just playing around, and legacy support. And that is another tangent.